src/server/verification/processor/provider/ZoomProvider.js

```diff
@@ -31,83 +38,130 @@ class ZoomProvider implements IEnrollmentProvider {
     onEnrollmentProcessing: (payload: IEnrollmentEventPayload) => void | Promise<void>,
     customLogger = null
   ): Promise<any> {
-    const { api, logger } = this
+    const { api, logger, _supportsFeature } = this
     const log = customLogger || logger
+    const { defaultMinimalMatchLevel, defaultSearchIndexName } = api
+    const { LivenessCheckFailed, SecurityCheckFailed, FacemapNotFound } = ZoomAPIError
+
     // send event to onEnrollmentProcessing
     const notifyProcessor = async eventPayload => onEnrollmentProcessing(eventPayload)
 
     // throws custom exception related to the predefined verification cases
     // e.g. livenes wasn't passed, duplicate found etc
     const throwCustomException = (customMessage, customResponse, zoomResponse = {}) => {
       const exception = new Error(customMessage)
+      // removing debug fields
+      let redactedResponse = omit(zoomResponse, 'callData', 'additionalSessionData', 'serverInfo')
+
+      // removing all large data (e.g. images , facemaps)
+      redactedResponse = omitBy(redactedResponse, (_, field) => field.endsWith('Base64'))
 
       exception.response = {
-        // removing all large data (e.g. images , facemaps)
-        ...pick(zoomResponse, 'code', 'subCode', 'message'),
+        ...redactedResponse,
         ...customResponse,
         isVerified: false
       }
 
       throw exception
     }
 
-    // 1. checking for duplicates
-    let duplicate
-    let faceSearchResponse
-    const { defaultMinimalMatchLevel } = api
+    let alreadyEnrolled
+    // checking is API supports delete enrollment endpoint
+    const isDisposeSupported = await _supportsFeature(ZoomAPIFeature.DisposeEnrollment)
 
+    // 1. checking if facescan already uploaded & enrolled
     try {
-      const { results, ...response } = await api.faceSearch(payload, defaultMinimalMatchLevel, customLogger)
+      await api.readEnrollment(enrollmentIdentifier, customLogger)
+      alreadyEnrolled = true
+    } catch (exception) {
+      // if something other that 'FacemapNotFound was thrown - re-throwing
+      if (FacemapNotFound !== exception.name) {
+        throw exception
+      }
 
-      faceSearchResponse = response
-      // excluding own enrollmentIdentifier
-      duplicate = results.find(
-        ({ enrollmentIdentifier: matchId }) => matchId.toLowerCase() !== enrollmentIdentifier.toLowerCase()
-      )
+      // otherwise, setting alreadyEnrolled to false
+      alreadyEnrolled = false
+    }
+
+    // 2. performing liveness check and storing facescan / audit trail images (if need)
+    let isLive = true
+
+    try {
+      const callArgs = [enrollmentIdentifier, payload, customLogger]
+      const enroll = () => api.submitEnrollment(...callArgs)
+
+      // if already enrolled, will call checkLiveness
+      // this is need because enroll doesn't re-checks facemap/images
+      // for liveness if identifier already enrolled
+      if (alreadyEnrolled) {
```

Comment:
what is this logic for alreadyEnrolled? did we have it before?

Reply:
@sirpy First of all please read my report about breaking changes in FV API here: https://app.zenhub.com/workspaces/gooddollar-5d50833888a83c6880d3e345/issues/gooddollar/gooddapp/2648#issuecomment-721209687
(between "Backend changes" and "I could propose" sentences)
I describing all the most significant changes and how i updated the FV flow according to them

a) Yes we had it before, but it worked some other way - we caught an enroll exception and if it's 'already exists' and there were no duplicates found - we resolving FV with 'alreadyEnrolled' flag
b) Now the logic is different because now the search requires the facemap will be enrolled and stored to the database. 
It doesn't works with the raw facemap / facescan as before. So enrollment should be the first operation and because of that we have some additional logic depending of alreadyEnrolled status:

1. check if alreadyEnrolled
2. if no - just do enroll
3. if yes:
3.1. check liveness. Enroll doesn't checks it if enrollment already exists so we couldn't call enroll and catch exception, we need to call liveness endpoint. Also this is needed to make sure that a) the real person passes re-verification b) re-enrollment will be successfull (if the server we're using allows us to remove the existing one)
3.2. if delete enrollment isn't - available - goto 4
3.3 if delete enrollment is supported - update facemap data:
3.3.1 delete previous enrollment from the 3d-db index
3.3.2 delete previous enrollment from the database
3.3.3 enroll the new facemap (it should be successfull as we're already confirmed liveness at pt. 3.1)
4. search for the duplicates using enrolled facemap (enrollmentIdentifier is passed now to the search endpoint isntead of the raw facemap as it was before
5. add enrollment identifier to the search index (without indexing enrollment doesn't appears in the search results)

Reply:
updated according to #290 and voice call discussion